R/tree.R

```diff
@@ -0,0 +1,314 @@
+        branch_hist = c(branch_hist, paste0(if(i == n_attributes) "last-", "attribute")),
+        attr_mode = TRUE # Let tree know this is an attribute
+      )
+    }
+  }
+}
+
+#' Build element or node label in tree
+#'
+#' These methods control how the value of a given node is printed. New methods
+#' can be added if support is needed for a novel class
+#'
+#' @inheritParams tree
+#'
+#' @export
+tree_label <- function(x, val_printer, class_printer, remove_newlines){
+  UseMethod("tree_label")
+}
+
+#' @export
+tree_label.function <- function(x, ...){
+  crayon::italic("function(){...}")
+}
+
+#' @export
+tree_label.environment <- function(x,...){
+  format.default(x)
+}
+
+#' @export
+tree_label.NULL <- function(x,...){
+  "<NULL>"
+}
+
+#' @export
+tree_label.character <- function(x, remove_newlines, ...){
+
+  # Get rid of new-line so they don't break tree flow
+  if(remove_newlines){
+    x <- gsub("\\n", replacement = " ", x = x, perl = TRUE)
+  }
+
+  # Shorten strings if needed
+  max_standalone_length <- 35
+  max_vec_length <- 15
+  max_length <- if(length(x) == 1) max_standalone_length else max_vec_length
+  x <- ifelse(
+    nchar(x) > max_length,
+    # Since we add an elipses we need to take a bit more than the max length
+    # off. The gsub adds elipses but also makes sure we dont awkwardly end on
+    # a space.
+    gsub(x = substr(x, start = 1, max_length - 3),
+         pattern = "\\s*$",
+         replacement = "...",
+         perl = TRUE),
+    x
+  )
+
+  tree_label.default(paste0("\"", x, "\""),...)
+}
+
+
+
+#' @export
+tree_label.default <- function(x, val_printer, class_printer, remove_newlines){
+
+  # There are a few psuedo-types that we want different printing behavior for.
+  # Since s3 methods cant differentiate between something like a single
+  # character and a vector of characters we use some logical branching here to
+  # try and use the best printing type for the passed value.
+  is_atomic_value <- is.atomic(x)
+  is_environment <- is.environment(x)
+
+  if(is_atomic_value) {
+
+    num_els <- length(x)
+    if(num_els > 1) {
+      # Atomic vectors are truncated to a max of 10 elements and printed inline
+      x <- as.character(x)
+      if(num_els > 10){
+        x <- head(x, 10)
+        x <- c(x, paste0("...(n = ", num_els, ")"))
+      }
+      x <- paste(x, collapse = ",")
+    }
+
+    # Single length atomics just go through unscathed
+    val_printer(x)
+
+  } else if(is.function(x)) {
+    # Lots of times function-like functions don't actually trigger the s3 method
+    # for function because they dont have function in their class-list. This
+    # catches those.
+    tree_label.function(x, class_printer, val_printer)
+  } else if(is.environment(x)) {
+    # Environments also tend to have the same trouble as functions. For instance
+    # the srcobject attached to a function's attributes is an environment but
+    # doesn't report as one to s3.
+    tree_label.environment(x)
```

Comment:
But `tree_label.environment` just calls `format.default`? I think it would be better to at least optionally recurse into environments.

Reply:
Agreed. I had put it off because it felt hard due to not knowing when to stop recusing. I think I have a good heuristic for stopping now and it's tested:
```R

if (is_environment(x)) {
  # If we're looking at an environment, we add its parent as child
  parent <- env_parent(x)

  already_seen_parent <- any(as.logical(lapply(counter_env$envs_seen, identical, parent)))
  # Stop recursion into environments when we get to either the calling
  # environment, the global environment, or an empty environment
  if (
    !(
      already_seen_parent |
      identical(parent, rlang::caller_env()) |
      identical(parent, rlang::empty_env()) |
      identical(parent, rlang::global_env())
    )
  ) {
    counter_env$envs_seen <- c(counter_env$envs_seen, parent)
    children <- c(children, parent = env_parent(x))
  }
}
```

```R
library(rlang)
ea <- env(d = 4, e = 5)
lobstr::tree(env(ea, a = 1, b = 2, c = 3))
#> <environment: 0x7f91ba1a4ab8>
#> ├─a: 1
#> ├─b: 2
#> ├─c: 3
#> └─parent: <environment: 0x7f91aa4d42e0>
#>   ├─d: 4
#>   └─e: 5
```

Reply:
I don't think it makes sense to show the parent -- for example, if you show the tree for `shiny:::.globals` (which is an environment), it'll show a bunch of environments that aren't relevant. I also think that if you do show parents, stopping at the global env or caller env can be misleading to users.

Reply:
To add a bit more: I think it's very uncommon when using environments to care what's in the parent env. That only really is relevant in closures. In the vast majority of cases, when an environment is used as a data structure, the parent env is not relevant, since `$` and `[[` indexing won't look in the parent.

Reply:
Yeah, I don't think you need to recurse into the parent either. 

Once you ignore the parent, you're less likely to encounter other special environments, but I'd suggest following the same heuristics as `obj_size()`: https://github.com/r-lib/lobstr/blob/master/src/size.cpp#L157-L159

Reply:
Recursing into parent was removed in ce4128a, also matches the forbidden environments from `obj_size()`.